### PR TITLE
fix: add requested filtering to next page url [GR-723]

### PR DESCRIPTION
**Pull request by ridingwolf**

### CODE REVIEW

File: src/Common/Infrastructure/BackendResponse.cs [hunk 1]
```diff
@@ -12,5 +15,48 @@ public BackendResponse(string content, string contentType, bool cameFromCache)
...
+            Content = Regex.Replace(
+                Content,
+                replaceData.Pattern,
+                replaceData.Replacement,
+                RegexOptions.IgnoreCase);
+        }
+
+        private string GetXmlPatternFor(string nextPageUrlValePattern)
+        {
+            return Regex.IsMatch(Content, "<feed xmlns=\"http://www.w3.org/2005/Atom\">")
```
Comment (CumpsD) [inline src/Common/Infrastructure/BackendResponse.cs]: Is Regex IsMatch faster than string Contains?

File: src/Common/Infrastructure/BackendResponse.cs [hunk 2]
```diff
@@ -12,5 +15,48 @@ public BackendResponse(string content, string contentType, bool cameFromCache)
...
+            const string numberPatternPlaceholder = "__NUMBER_PATTERN_PLACEHOLDER_THAT_WONT_BE_REGEX_ESCAPED__";
+            var nextPageUrlOptionWithPlaceholders = string.Format(nextPageUrlOption, numberPatternPlaceholder, numberPatternPlaceholder);
+            var nextPageUrlValePattern = Regex
+                .Escape(nextPageUrlOptionWithPlaceholders)
+                .Replace(numberPatternPlaceholder, "\\d+");
+
+            var parameters = requestQuery
+                .Aggregate("", (result, filterQueryParameter) => $"{result}&{filterQueryParameter.Key}={filterQueryParameter.Value}");
+
+            var isXml = ContentType.ToLower().Contains("xml");
```
Comment (CumpsD) [inline src/Common/Infrastructure/BackendResponse.cs]: Can we add a new class `ContentType` like we did with `AcceptType`? https://github.com/Informatievlaanderen/public-api/blob/134493d8f13e8e8618e87ffa614f23999337414b/src/Common/Infrastructure/AcceptType.cs

Then we can get rid of the stringy contenttype in here as well

File: src/Common/Infrastructure/NonPagedQueryCollection.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+            bool IsNotPageParameter(KeyValuePair<string, StringValues> keyValuePair) =>
+                false == PageQueryParameters.Contains(keyValuePair.Key, StringComparer.InvariantCultureIgnoreCase);
+
+            return query
+                .Where(queryParameter => queryParameter.Value.Count > 0)
+                .Where(IsNotPageParameter)
+                .ToDictionary(queryParameter => queryParameter.Key, queryParameter => queryParameter.Value);
+        }
+
+        public bool IsEmpty => Count == 0;
```
Comment (CumpsD) [inline src/Common/Infrastructure/NonPagedQueryCollection.cs]: can you put this prop above the ctor

File: src/Common/Infrastructure/RegistryApiController.cs [hunk 4]
```diff
@@ -54,5 +57,23 @@ public abstract class RegistryApiController<T> : ApiController<T>
...
+                => !string.IsNullOrWhiteSpace(parameter.Value.ToString());
+
+            return Request.Query.Count == 0
+                ? keyBaseValue
+                : Request
+                    .Query
+                    .Where(ParameterHasValue)
+                    .OrderBy(queryParameter => queryParameter.Key)
+                    .Aggregate(keyBaseValue, (key, queryParameter) => $"{key}-{queryParameter.Key}:{queryParameter.Value}")
+                    .Replace(":-", ":");
```
Comment (CumpsD) [inline src/Common/Infrastructure/RegistryApiController.cs]: I think this needs to be removed, a queryparameter with a `-` in front means descending, so it is really a different cachekey

File: src/Public.Api/Address/AddressController-List.cs [hunk 5]
```diff
@@ -158,13 +165,13 @@ void HandleBadRequest(HttpStatusCode statusCode)
...
 
-            var cacheKey = $"legacy/address-list:{offset}-{limit}-{taal}";
+            var cacheKey = CreateCacheKeyForRequestQuery($"legacy/address-list:{taal}");
 
             var value = await (CacheToggle.FeatureEnabled
                 ? GetFromCacheThenFromBackendAsync(format, BackendRequest, cacheKey, Request.GetTypedHeaders(), HandleBadRequest, cancellationToken)
                 : GetFromBackendAsync(format, BackendRequest, Request.GetTypedHeaders(), HandleBadRequest, cancellationToken));
 
-            return new BackendResponseResult(value);
+            return  BackendListResponseResult.Create(value, Request.Query, responseOptions.Value.VolgendeUrl);
```
Comment (CumpsD) [inline src/Public.Api/Address/AddressController-List.cs]: double space after return?

File: src/Common/Infrastructure/BackendResponse.cs (hunk 1 above)
Comment (ridingwolf) [reply src/Common/Infrastructure/BackendResponse.cs]: nope, forgot to replace after changing the patteren to a regular string

**Comment by CumpsD** [comment]
You'll probably have to pull latest master and rebase too

File: src/Common/Infrastructure/BackendResponse.cs (hunk 1 above)
Comment (CumpsD) [reply src/Common/Infrastructure/BackendResponse.cs]: for reference sake later on: https://theburningmonk.com/2012/05/performance-test-string-contains-vs-string-indexof-vs-regex-ismatch/ :)

**Comment by CumpsD** [comment]
general remark for the PR: have a look into compiled regex's too if we can re-use any

**Comment by CumpsD** [comment]
nevermind my last remark, I dont see any cases where they can be compiled

File: src/Common/Infrastructure/BackendResponse.cs (hunk 1 above)
Comment (ridingwolf) [reply src/Common/Infrastructure/BackendResponse.cs]: Does not differentiate between a lazy/non-lazy regex though ;-P
Anyway, replacing the Regex.IsMatch

File: src/Common/Infrastructure/BackendResponse.cs (hunk 2 above)
Comment (ridingwolf) [reply src/Common/Infrastructure/BackendResponse.cs]: ContentType is just response property, too much overhead to create type

File: src/Common/Infrastructure/RegistryApiController.cs (hunk 4 above)
Comment (ridingwolf) [reply src/Common/Infrastructure/RegistryApiController.cs]: replace ":-", ":" is only there to clean up keys like 
"legacy/parcel-list:-limit:100-offset:0" 
grouped parameter name+value for readability to "legacy/parcel-list:(limit:100)-(offset:0)"

**Comment by ridingwolf** [comment]
Fixed what made sense and did slight readablilty refactor on CacheKey & Repalce Pattern/ReplaceValue

File: src/Common/Infrastructure/RegistryApiController.cs [hunk 6]
```diff
@@ -72,8 +72,8 @@ bool ParameterHasValue(KeyValuePair<string, StringValues> parameter)
                     .Query
                     .Where(ParameterHasValue)
                     .OrderBy(queryParameter => queryParameter.Key)
-                    .Aggregate(keyBaseValue, (key, queryParameter) => $"{key}-{queryParameter.Key}:{queryParameter.Value}")
-                    .Replace(":-", ":");
+                    .Aggregate(keyBaseValue, (key, queryParameter) => $"{key}-({queryParameter.Key}:{queryParameter.Value})")
+                    .Replace(":-(", ":(");
```
Comment (CumpsD) [inline src/Common/Infrastructure/RegistryApiController.cs]: :( :-(

sad haha :p

**Pull request merged**